### ISSUE TITLE
Mimir Alertmanager failing to load config as a global opsgenie_api_key is not specified

### ISSUE THREAD

**Issue by amrit-canva** [grafana/mimir#8723]
#### Describe the bug
Since upgrading to Mimir `2.13.0`, Mimir Alertmanager is not loading our existing config and is giving this error:

```
grafana-mimir-alertmanager-0 alertmanager ts=2024-07-15T03:18:41.973420345Z caller=multitenant.go:653 level=warn component=MultiTenantAlertmanager msg="error applying config" err="unable to load fallback configuration for anonymous: no global OpsGenie API
 Key set"
```

#### To Reproduce

Update the alertmanager fallback config to include an Opsgenie receiver and do not set `opsgenie_api_key` in the global config.

Alertmanager won't come up and in logs an error will be produced indicating that global opsgenie API key is not set.

#### Expected behavior

`opsgenie_api_key` is not a mandatory config parameter. In our setup we use `opsgenie_api_key_file` and specify this at per receiver level, instead of in the global config.

So I expect that the presence of `opsgenie_api_key` in global config section must not be manadatory.

#### Environment

- Infrastructure: Kubernetes
- Deployment tool: helm

#### Additional Context

Seems like this behaviour is introduced through this check in `grafana/alerting` repo: https://github.com/grafana/alerting/blob/main/definition/compat.go#L115-L119

**Comment by antonio-tolentino** [comment]
I'm facing the same issue with the introduction of this new Grafana definitions validation in Mimir 2.13. From Alertmanager perspective **api_key_file** field for Opsgenie is still entirely supported.

Questions:
- Why the Opsgenie receiver field **api_key** is being enforced like this ? 
- Is the Opsgenie receiver field **api_key_file** being deprecated for Mimir Alermanager?
- Should I expect that local storage for Mimir alertmanager to be deprecated too?
- Is there another recommend way to inject **api_key** to avoid this be flat in the file?

**Comment by daanschipper** [comment]
Encountered the same issue `no global Slack API URL set` when using `api_url_file` in a `slack_configs` receiver. Workaround is to specify a fake url under `global.slack_api_url`. The value has to be a valid http url. The global value is ignored as it is only a default value when no url is specified under the receiver.

**Comment by anthonynashduco** [comment]
Hitting this wile trying to set `global.slack_api_url_file`. Got a valid webhook url in there but it's just failing to start complaining that `global.slack_api_url` isn't set. Is there an accepted way of _not_ having to set plain text secrets in the config? One that isn't setting the whole config in a secret...

**Comment by yuri-tceretian** [comment]
I do not think this should have ever worked in Mimir. API prohibits such fields. 
https://github.com/grafana/mimir/blob/a5f6bc75e858f2f7ede4b68bd692ed9b4f99193d/pkg/alertmanager/api.go#L493-L507

The fact that it worked for fallback configuration is rather an oversight. I would recommend migrating away from `_file` settings

**Comment by yuri-tceretian** [comment]
> mimirtool alertmanager verify alertmanager.yaml did not give any validation error, the [changelog](https://github.com/grafana/mimir/blob/main/CHANGELOG.md?rgh-link-date=2024-10-29T10%3A37%3A51.000Z) mention anything

This is a good point. I think mimirtool validation could be improved.

**Comment by amrit-canva** [comment]
> I do not think this should have ever worked in Mimir. API prohibits such fields.
> 
> [mimir/pkg/alertmanager/api.go](https://github.com/grafana/mimir/blob/a5f6bc75e858f2f7ede4b68bd692ed9b4f99193d/pkg/alertmanager/api.go#L493-L507)
> 
> Lines 493 to 507 in [a5f6bc7](/grafana/mimir/commit/a5f6bc75e858f2f7ede4b68bd692ed9b4f99193d)
> 
>  func validateGlobalConfig(cfg config.GlobalConfig) error { 
>  	if cfg.SlackAPIURLFile != "" { 
>  		return errSlackAPIURLFileNotAllowed 
>  	} 
>  	if cfg.OpsGenieAPIKeyFile != "" { 
>  		return errOpsGenieAPIKeyFileFileNotAllowed 
>  	} 
>  	if cfg.SMTPAuthPasswordFile != "" { 
>  		return errPasswordFileNotAllowed 
>  	} 
>  	if cfg.VictorOpsAPIKeyFile != "" { 
>  		return errVictorOpsAPIKeyFileNotAllowed 
>  	} 
>  	return nil 
>  } 
> The fact that it worked for fallback configuration is rather an oversight. I would recommend migrating away from `_file` settings

Okay so in case of alertmanager it seems that Mimir does not support the Prometheus alertmanager config file completely without changes? If that is the case, I reckon it will be better to mention it in the Mimir alertmanager documentation https://grafana.com/docs/mimir/latest/references/architecture/components/alertmanager/#tenant-configurations

<img width="830" height="175" alt="Image" src="https://github.com/user-attachments/assets/1cf55fcd-a3ee-4a2e-bde2-7839e4efcd77" />

**Comment by cloudymax** [comment]
So what is the appropriate way to have alert-manager get the `slack_api_url` or `opsgenie_api_url` from an existing secret or file?

**Comment by naveenkumarsp** [comment]
@cloudymax  -  you may need to use dummy vaule in global config and use the actual value in 
```
global:
  opsgenie_api_key: "dummy"  # Placeholder to pass validation

receivers:
  - name: opsgenie
    opsgenie_configs:
      - api_key_file: /path/to/api/key/file
        responders:
          - name: your-team
            type: team
```

**Comment by artemlive** [comment]
> [@cloudymax](https://github.com/cloudymax) - you may need to use dummy vaule in global config and use the actual value in
> 
> ```
> global:
>   opsgenie_api_key: "dummy"  # Placeholder to pass validation
> 
> receivers:
>   - name: opsgenie
>     opsgenie_configs:
>       - api_key_file: /path/to/api/key/file
>         responders:
>           - name: your-team
>             type: team
> ```

This won't help, since it will use the specified dummy key, because if we don't specify the per-receiver APIKey, it will use the global one:
https://github.com/grafana/mimir/blob/ded396d5ac1db5a1d179077fc12cfbdc644d4d1a/vendor/github.com/grafana/alerting/definition/compat.go#L117-L122

Later, it won't use the file, since we specified the APIKey explicitly:
https://github.com/grafana/mimir/blob/ded396d5ac1db5a1d179077fc12cfbdc644d4d1a/vendor/github.com/prometheus/alertmanager/notify/opsgenie/opsgenie.go#L277-L286

**Comment by artemlive** [comment]
> So what is the appropriate way to have alert-manager get the `slack_api_url` or `opsgenie_api_url` from an existing secret or file?

https://grafana.com/docs/grafana-cloud/send-data/alloy/reference/components/mimir/mimir.alerts.kubernetes/#usage
I'd suggest you use Alloy's new feature. It can send the AlertmanagerConfig directly to Mimir. And you can also use secrets to store your API keys.
